src/librustc_mir/transform/inline.rs

```diff
@@ -571,11 +581,47 @@ impl<'a, 'tcx> Inliner<'a, 'tcx> {
 
             let stmt = Statement {
                 source_info: callsite.location,
-                kind: StatementKind::Assign(arg_tmp.clone(), arg)
+                kind: StatementKind::Assign(arg_tmp.clone(), arg),
             };
             caller_mir[callsite.bb].statements.push(stmt);
             Operand::Consume(arg_tmp)
-        }).collect()
+        }
+
+        let tcx = self.tcx;
+
+        // A closure is passed its self-type and a tuple like `(arg1, arg2, ...)`,
+        // hence mappings to tuple fields are needed.
+        if tcx.def_key(callsite.callee).disambiguated_data.data == DefPathData::ClosureExpr {
```

Comment:
Nit: let's do `if let DefPathData::ClosureExpr = tcx.def_key(callsite.callee).disambiguated_data.data`

Reply:
better yet, let's add a `is_closure(def_id: DefId)` helper to the tcx (and call it from `closure_base_def_id`)